File: 3316348dbadfb7b11c7c2ea235949419e23f9fa898ad2c198f999617912a9925.py
# 3316348dbadfb7b11c7c2ea235949419e23f9fa898ad2c198f999617912a9925.py —— read 插件（变量名在左 → 压栈，顶层执行）
# 语义：从 var 区反向找同名变量，把值拷贝到 stk（对应 C：memcmp 匹配后 memcpy(stk_off, vptr, vsize)）
import struct
import vmstate

name = payload.encode() if isinstance(payload, str) else payload
v = vmstate.var_off
while v > 0:
    nsize = struct.unpack_from("<I", vmstate.var, v-16)[0]
    vptr  = struct.unpack_from("<Q", vmstate.var, v-12)[0]
    vsize = struct.unpack_from("<I", vmstate.var, v-4)[0]
    if vmstate.var[v-16-nsize:v-16] == name:   # 名字匹配
        vmstate.push(bytes(vmstate.stk[vptr:vptr+vsize]))  # 值在 stk（vptr 是 stk 偏移）
        break
    v -= 16 + nsize

File: c/plugins/3316348dbadfb7b11c7c2ea235949419e23f9fa898ad2c198f999617912a9925.c
#include "../vm.h"
#include <stdlib.h>
#include <string.h>

void plugin_run(VM *vm, const uint8_t *payload, uint32_t plen) {
    uint32_t v = vm->var_off;
    while (v > 0) {
        uint32_t nsize, vsize; uint64_t vptr;
        memcpy(&nsize, vm->var + v - 16, 4);
        memcpy(&vptr, vm->var + v - 12, 8);
        memcpy(&vsize, vm->var + v - 4, 4);
        if (nsize == plen && memcmp(vm->var + v - 16 - nsize, payload, plen) == 0) {
            vm->cb_push(vm, vm->stk + (size_t)vptr, vsize);
            break;
        }
        v -= 16 + nsize;
    }
    vm->cb_run_next(vm);
}
